docs/src/dev/principles/unsafe.md

```diff
@@ -0,0 +1,243 @@
+/// The caller must ensure that `index` is within the bounds of the collection.
+unsafe fn get_element_unchecked<T>(slice: &[T], index: usize) -> &T {
+  // SAFETY: Caller must ensure that index < slice.len()
+  unsafe { slice.get_unchecked(index) }
+}
+```
+
+In this case, the invariant is that `index` is within the bounds of `slice`. It is up to the programmer to decide how
+this interface should be defined and whether the function is unsafe. For example, the function could have been written
+as:
+
+```rust
+/// Returns an Option containing a reference to the element at the specified index without performing bounds checking.
+///
+/// # Parameters
+/// - `index`: The position of the element to retrieve.
+///
+/// # Returns
+/// An Option containing a reference to the element at the given index.
+///
+/// # Safety
+/// Calling this function with an out-of-bounds index is undefined behavior.
+/// The caller must ensure that `index` is within the bounds of the collection.
+fn get_element_unchecked<T>(slice: &[T], index: usize) -> Option<&T> {
+  if index >= slice.len() {
+    return None;
+  }
+  // SAFETY: Caller must ensure that index < slice.len()
+  unsafe { Some(slice.get_unchecked(index)) }
+}
+```
+
+In this version, the function is no longer unsafe, despite containing an unsafe code block within it, because it no
+longer has an invariant; the function itself is able to manage whether its inputs are valid.
+
+Whenever possible, programmers should write safe functions that validate all preconditions, postconditions, and
+invariants. Of course, this is not always possible, in which case unsafe functions are called for.
+
+For hardware safety, we don't have invariants, necessarily. For example, in writing a system register, there is no
+invariant that must hold true for that operation. There may be preconditions or postconditions, but not an invariant.
+
+One simple example of hardware safety is the `invld` instruction on x86_64 that invalidates caches without writing them
+back. This operation can be very dangerous if not called at a proper time, but nothing about the software can change
+that; there is no validation that can be done to indicate this is a safe time to call this. In Patina, this is
+hardware safety and does not warrant an unsafe function.
+
+We may define a wrapper function for it as:
+
+```rust
+fn invalidate_caches() {
+  // SAFETY: This is an architecturally defined way to invalidate caches
+  unsafe { asm!("invld") }
+}
+```
+
+Software safety could be introduced, say in the below example:
+
+```rust
+// SAFETY: Caller must ensure that `ptr` is a valid u8 pointer
+unsafe fn invalidate_caches(ptr: *mut u8) {
+  // Better set this ptr to 4 first!
+  // SAFETY: See function comment, caller assumes responsibility
+  unsafe {ptr.write(4);}
+
+  // SAFETY: This is an architecturally defined way to invalidate caches
+  unsafe { asm!("invld") } 
+}
+```
+
+However, the function could be made safe again by validating assumptions:

```

Comment:
In the following example though, it's not declared safe (and shouldn't be because its still not safe to dereference and arbitrary NonNull). Perhaps this was meant to be changed?